from pathlib import Path
import json

BAMS = {
    "ont": "https://giab-data.s3.amazonaws.com/XY_evaluation/11_16_22_R1041_HG002_UL_Kit14_260_Guppy_6.3.8_sup.pass.lt100kb_sorted_haplotagged.bam",
    "hifi": "https://ftp-trace.ncbi.nlm.nih.gov/ReferenceSamples/giab/data/AshkenazimTrio/HG002_NA24385_son/PacBio_CCS_15kb_20kb_chemistry2/GRCh38/HG002.SequelII.merged_15kb_20kb.pbmm2.GRCh38.haplotag.10x.bam",
}

IDS = list(map(str, [81, 82, 84, 85, 89]))

REGIONS = {
    "IGK": "chr2",
    "TRG": "chr7:0-100000000",
    "TRB": "chr7:100000000",
    "TRA": "chr14:0-80000000",
    "IGH": "chr14:80000000",
    "IGL": "chr22",
}


# wildcard_constraints:
#     gene=f"({'|'.join(GENES)})",


IMGT_URL = "https://www.imgt.org/download/GENE-DB/IMGTGENEDB-ReferenceSequences.fasta-nt-WithoutGaps-F+ORF+allP"

ALIGN_PATH = Path("results") / "immuno_alignments" / "{bamid}"


rule intersect_bam:
    input:
        "static/immuno.bed",
    output:
        ALIGN_PATH / "raw.bam",
    params:
        url=lambda w: BAMS[w.bamid],
    conda:
        "envs/bedtools.yml"
    shell:
        """
        curl -LqSs --fail {params.url} | \
        bedtools intersect -abam stdin -b {input} -ubam -u > {output}
        """


rule download_imgt_goodies:
    output:
        "resources/imgt_human.fa",
    params:
        url=IMGT_URL,
    conda:
        "envs/master.yml"
    shell:
        """
        curl -SsLq --fail {params.url} | \
        seqkit grep -r -n -p '.*Homo sapiens\|.*\|(V|D|J)-REGION.*' > {output}
        """


rule split_igmt_goodies:
    input:
        rules.download_imgt_goodies.output,
    output:
        "results/igmt/{locus}.fa",
    conda:
        "envs/master.yml"
    shell:
        """
        seqkit grep -r -n -p '.*|{wildcards.locus}.*?|.*' {input} > {output}
        """


rule parse_igmt_headers:
    input:
        rules.download_imgt_goodies.output,
    output:
        "results/igmt/headers.tsv",
    shell:
        """
        cat {input} | \
        grep '>' | \
        sed 's/>//' | \
        sed 's/|$//' |
        sed 's/|/\t/g' > {output}
        """


rule index_bam:
    input:
        rules.intersect_bam.output,
    output:
        rules.intersect_bam.output[0] + ".bai",
    conda:
        "envs/master.yml"
    shell:
        """
        samtools index {input}
        """


rule split_bam:
    input:
        _idx=rules.index_bam.output,
        bam=rules.intersect_bam.output,
    output:
        ALIGN_PATH / "split" / "{locus}.fastq",
    params:
        region=lambda w: REGIONS[w.locus],
    conda:
        "envs/master.yml"
    log:
        "log/{bamid}/split/{locus}.log",
    shell:
        """
        samtools view -h {input.bam} {params.region} | \
        samtools fastq > {output} 2> {log}
        """


rule split_fastq:
    input:
        bam=rules.split_bam.output[0],
        alleles=rules.split_igmt_goodies.output[0],
    output:
        fasta=directory(ALIGN_PATH / "split" / "{locus}" / "fasta"),
        sam=directory(ALIGN_PATH / "split" / "{locus}" / "aligned"),
        merged=ALIGN_PATH / "split" / "{locus}" / "merged.bam"
    log:
        index="log/{bamid}/index/{locus}.log",
        align="log/{bamid}/align/{locus}.log",
    threads: 10
    conda:
        "envs/master.yml"
    script:
        "scripts/split_fastq.py"


# rule index_fasta:
#     input:
#         str(Path(rules.split_fastq.output[0]) / "{i}.fasta"),
#     output:
#         str(Path(rules.split_fastq.output[0]) / "{i}.fasta.bwt"),
#     log:
#         "log/{bamid}/index/{locus}/{i}.log",
#     conda:
#         "envs/master.yml"
#     shell:
#         """
#         bwa index {input} 2> {log}
#         """


# rule align_alleles:
#     input:
#         _idx=rules.index_fasta.output,
#         template=rules.index_fasta.input,
#         alleles=rules.split_igmt_goodies.output,
#     output:
#         ALIGN_PATH / "split" / "{locus}" / "aligned" / "{i}.bam",
#     log:
#         "log/{bamid}/align/{locus}/{i}.log",
#     conda:
#         "envs/master.yml"
#     shell:
#         """
#         bwa mem -a {input.template} {input.alleles} 2> {log} | samtools view -b > {output}
#         """


# def fasta_input(wildcards):
#     o = checkpoints.split_fastq.get(**wildcards).output[0]
#     return expand(
#         Path(rules.split_fastq.output.sam) / "{i}.bam",
#         locus=wildcards.locus,
#         bamid=wildcards.bamid,
#         i=glob_wildcards(os.path.join(o, "{i}.fasta")).i,
#     )


rule cat_alignments:
    input:
        rules.split_fastq.output.merged,
    output:
        sam=ALIGN_PATH / "split" / "{locus}" / "all.sam",
        tags=ALIGN_PATH / "split" / "{locus}" / "all.tags",
    conda:
        "envs/master.yml"
    shell:
        """
        samtools view -F 0x4 -F 0x100 {input} | \
        grep -v "SA:" | \
        tee >(cut -f -11 > {output.sam}) | \
        cut -f 12- > {output.tags}
        """


rule all:
    input:
        expand(rules.cat_alignments.output, locus=REGIONS, bamid=BAMS),
